### PR TITLE
Specify non-random location for filesystem's temporary cache

### PR DESCRIPTION
**Pull request by oandreeva-nv**
Main ask for [DLIS-5340](https://jirasw.nvidia.com/browse/DLIS-5340)

Previous PR: https://github.com/triton-inference-server/core/pull/254 needs some debugging.

This PR will only read env variable directory (if specified) and create temporary folder there.

### CODE REVIEW

File: src/filesystem/implementations/gcs.h [hunk 1]
```diff
@@ -380,9 +381,12 @@ GCSFileSystem::LocalizePath(
         "GCS file localization not yet implemented " + path);
   }
 
+  // Create a local directory for GCS model store.
+  std::string env_mount_dir =
+      GetEnvironmentVariableOrDefault("TRITON_GCS_MOUNT_DIRECTORY", "/tmp");
```
Comment (rmccorm4) [inline src/filesystem/implementations/gcs.h:386]: Can `"/tmp"` be some `common.h` constant like `constexpr char kDefaultMountDirectory[] = "/tmp"` or whatever name makes most sense?

File: src/filesystem/api.h [hunk 2]
```diff
@@ -198,6 +198,15 @@ Status MakeDirectory(const std::string& dir, const bool recursive);
...
 Status MakeTemporaryDirectory(const FileSystemType type, std::string* temp_dir);
 
+/// Create a temporary directory of the specified filesystem type
+/// in a provided location.
+/// \param type The type of the filesystem.
+/// \param  dir_path The specified path.
+/// \param temp_dir Returns the path to the temporary directory.
+/// \return Error status
+Status MakeTemporaryDirectory(
+    const FileSystemType type, std::string& dir_path, std::string* temp_dir);
```
Comment (rmccorm4) [inline src/filesystem/api.h:208]: I think `dir_path` is defined as a reference here `std::string& dir_path`, but a copy `std::string` everywhere else. Should it be copy here? It should be consistent.

Comment (oandreeva-nv) [reply src/filesystem/api.h:208]: Yes, correct. Something strange happening with my local repo

File: src/filesystem/implementations/local.h [hunk 3]
```diff
@@ -314,7 +316,10 @@ LocalFileSystem::MakeTemporaryDirectory(std::string* temp_dir)
         "Failed to create local temp folder: " + *temp_dir);
   }
 #else
-  std::string folder_template = "/tmp/folderXXXXXX";
+  if (dir_path.empty()) {
+    dir_path = "/tmp";
```
Comment (rmccorm4) [inline src/filesystem/implementations/local.h:320]: use constant

**Review by rmccorm4 — APPROVED** [review]

File: src/filesystem/api.h [hunk 4]
```diff
@@ -198,6 +198,15 @@ Status MakeDirectory(const std::string& dir, const bool recursive);
...
 Status MakeTemporaryDirectory(const FileSystemType type, std::string* temp_dir);
 
+/// Create a temporary directory of the specified filesystem type
+/// in a provided location.
+/// \param type The type of the filesystem.
+/// \param  dir_path The specified path.
+/// \param temp_dir Returns the path to the temporary directory.
+/// \return Error status
+Status MakeTemporaryDirectory(
+    const FileSystemType type, std::string dir_path, std::string* temp_dir);
```
Comment (kthui) [inline src/filesystem/api.h:208]: A small thing, I think the `dir_path` can be declared as `const std::string& dir_path` here, and leave the `LocalFileSystem::MakeTemporaryDirectory(std::string dir_path, ...)` as it is. Then, we can save a copy of `dir_path` when passing the variable from the caller to `MakeTemporaryDirectory()` here, and only copy it once when passing `dir_path` from here to `LocalFileSystem::MakeTemporaryDirectory()`.

Comment (kthui) [inline src/filesystem/api.h:208]:
```suggestion
    const FileSystemType type, const std::string& dir_path, std::string* temp_dir);
```

File: src/filesystem/api.cc [hunk 5]
```diff
@@ -619,7 +619,16 @@ MakeTemporaryDirectory(const FileSystemType type, std::string* temp_dir)
 {
   std::shared_ptr<FileSystem> fs;
   RETURN_IF_ERROR(fsm_.GetFileSystem(type, fs));
-  return fs->MakeTemporaryDirectory(temp_dir);
+  return fs->MakeTemporaryDirectory(kDefaultMountDirectory, temp_dir);
+}
+
+Status
+MakeTemporaryDirectory(
+    const FileSystemType type, std::string dir_path, std::string* temp_dir)
```
Comment (kthui) [inline src/filesystem/api.cc:627]:
```suggestion
    const FileSystemType type, const std::string& dir_path, std::string* temp_dir)
```

**Review by kthui — APPROVED** [review]

File: src/filesystem/api.h (hunk 4 above)
Comment (oandreeva-nv) [reply src/filesystem/api.h:208]: I'll make sure to check and introduce this in a follow-up pr with the bug fix.   Thanks for suggestion!

**Pull request merged**